metrics/repl.lag/lag.go

```diff
@@ -115,70 +130,122 @@ func (c *Lag) Help() blip.CollectorHelp {
 	}
 }
 
+// Prepare prepares lag collectors for all levels in the plan that contain the "repl.lag" domain
 func (c *Lag) Prepare(ctx context.Context, plan blip.Plan) (func(), error) {
 LEVEL:
-	for _, level := range plan.Levels {
+	for levelName, level := range plan.Levels {
 		dom, ok := level.Collect[DOMAIN]
 		if !ok {
 			continue LEVEL // not collected in this level
 		}
+		cleanup, err := c.prepareLevel(levelName, plan.MonitorId, plan.Name, dom.Options)
+		if err != nil {
+			return cleanup, err
+		}
+	}
+	return nil, nil
+}
 
-		table := dom.Options[OPT_TABLE]
-		if table == "" {
-			table = blip.DEFAULT_HEARTBEAT_TABLE
+func (c *Lag) Collect(ctx context.Context, levelName string) ([]blip.MetricValue, error) {
+	switch c.lagSourceIn[levelName] {
+	case LAG_SOURCE_BLIP:
+		return c.collectBlipHeartBeatLag(ctx, levelName)
+	case LAG_SOURCE_PFS:
+		return c.collectLagFromPFS(ctx, levelName)
+	}
+
+	panic(fmt.Sprintf("invalid lag source in Collect %s", c.lagSourceIn[levelName]))
+}
+
+// //////////////////////////////////////////////////////////////////////////
+// Internal methods
+// //////////////////////////////////////////////////////////////////////////
+
+func (c *Lag) prepareLevel(levelName string, monitorID string, monitorName string, options map[string]string) (func(), error) {
+	c.dropNotAReplica[levelName] = !blip.Bool(options[OPT_REPORT_NOT_A_REPLICA])
+	c.replCheck = sqlutil.CleanObjectName(options[OPT_REPL_CHECK]) // @todo sanitize better
+
+	if source, ok := options[OPT_LAG_SOURCE]; ok {
+		if len(source) > 0 && source != "auto" {
+			switch source {
+			case LAG_SOURCE_PFS:
+				return nil, c.preparePFS(levelName)
+			case LAG_SOURCE_BLIP:
+				return c.prepareBlipHeartbeatLag(levelName, monitorID, monitorName, options)
+			default:
+				return nil, fmt.Errorf("invalid lag source: %s; valid values: auto, pfs, blip", source)
+			}
 		}
+	}
+
+	// -------------------------------------------------------------------------
+	// Auto source (default)
+	// -------------------------------------------------------------------------
+	var err error
+	if err = c.preparePFS(levelName); err == nil {
+		return nil, nil
+	}
+
+	if cleanup, err := c.prepareBlipHeartbeatLag(levelName, monitorID, monitorName, options); err == nil {
+		return cleanup, nil
+	}
+
+	return nil, fmt.Errorf("auto lag source failed, last error: %s", err)
+}
 
-		c.dropNoHeartbeat[level.Name] = !blip.Bool(dom.Options[OPT_REPORT_NO_HEARTBEAT])
-		c.dropNotAReplica[level.Name] = !blip.Bool(dom.Options[OPT_REPORT_NOT_A_REPLICA])
+func (c *Lag) prepareBlipHeartbeatLag(levelName string, monitorID string, planName string, options map[string]string) (func(), error) {
+	if c.lagReader == nil {
+		c.dropNoHeartbeat[levelName] = !blip.Bool(options[OPT_REPORT_NO_HEARTBEAT])
 
+		table := options[OPT_HEARTBEAT_TABLE]
```

Comment:
Nit: If you're testing for map membership, it's more idiomatic to do this:
```
table, ok := options[OPT_HEARTBEAT_TABLE]
if !ok {
     table = blip.DEFAULT_HEARTBEAT_TABLE
}
```

Reply:
In this instance, I want to set it to default if it's not set or it's **intentionally** set to "" (empty string), so I think this is cleaner as I don't need to distinguish between empty or unset key.